### PR TITLE
Fix payment method showing None for paid subscribers

### PR DESCRIPTION
**Pull request by chiedoclaude**
## Summary
- The `has_payment_method` field in the billing API was only checking the local `payment_methods` table (used for agent-initiated purchases)
- Users on pay-as-you-go plans have a Stripe payment method from checkout but may not have added a local card for agent transactions
- This caused the Plan Details section to incorrectly show "Payment Method: None" even for paying subscribers
- Fixed by also considering `stripe_subscription_id` — if set, the user completed Stripe checkout and has a payment method on file

### Claude Code
- [x] Backend builds cleanly (`go build ./...`)
- [x] All API tests pass (`go test ./api/...`)
- [x] All DB tests pass (`go test ./db/...`)
- [x] All frontend tests pass (851/851)

### OpenClaw
- [ ] Verify on app.permissionslip.dev that pay-as-you-go users see "On file" instead of "None"
- [ ] Verify free-tier users without payment methods still see "None"
- [ ] Verify users with local payment methods (for agent purchases) still see "On file"

https://claude.ai/code/session_0199w7S523xCcft15d17Pxzc

### CODE REVIEW

File: api/billing.go [hunk 1]
```diff
@@ -236,12 +236,16 @@ func handleGetBillingPlan(deps *Deps) http.HandlerFunc {
 			Subscription: newBillingSubscription(sub),
 		}
 
-		// Check actual payment methods rather than relying on Stripe customer ID.
+		// A user "has a payment method" if they have local payment methods
+		// (for agent-initiated purchases) OR an active Stripe subscription
+		// (which means they provided Stripe with a payment method at checkout).
+		hasStripeSubscription := sub.StripeSubscriptionID != nil
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/billing.go]: Good catch. Added `&& sub.PlanID != db.PlanFree` guard to both locations and a test for the downgraded user scenario (`fd57738`).

File: api/billing.go [hunk 2]
```diff
@@ -304,12 +308,16 @@ func handleGetSubscription(deps *Deps) http.HandlerFunc {
 			PlanLimits:          newPlanLimits(&sub.Plan),
 		}
 
-		// Check actual payment methods rather than relying on Stripe customer ID.
+		// A user "has a payment method" if they have local payment methods
+		// (for agent-initiated purchases) OR an active Stripe subscription
+		// (which means they provided Stripe with a payment method at checkout).
+		hasStripeSubscription := sub.StripeSubscriptionID != nil
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/billing.go]: Fixed in the same commit — both locations now have the plan guard (`fd57738`).

File: api/billing.go [hunk 3]
```diff
@@ -236,12 +236,16 @@ func handleGetBillingPlan(deps *Deps) http.HandlerFunc {
...
 		if pmCount, err := db.CountPaymentMethodsByUser(r.Context(), deps.DB, profile.ID); err != nil {
 			log.Printf("[%s] GetBillingPlan: count payment methods: %v", TraceID(r.Context()), err)
 			CaptureError(r.Context(), err)
+			resp.Subscription.HasPaymentMethod = hasStripeSubscription
 		} else {
-			resp.Subscription.HasPaymentMethod = pmCount > 0
+			resp.Subscription.HasPaymentMethod = pmCount > 0 || hasStripeSubscription
 		}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/billing.go:251]: Added `TestGetBillingPlan_HasPaymentMethodFalseAfterDowngrade` — inserts a free-plan user with a stale `stripe_subscription_id` and asserts `has_payment_method=false` (`fd57738`).

File: frontend/src/pages/billing/PlanDetailsCard.tsx [hunk 4]
```diff
@@ -41,6 +41,30 @@ function CostEstimate() {
...
+  const { openPortal, isLoading } = useBillingPortal();
+
+  return (
+    <div className="flex items-center justify-between">
+      <span className="text-sm font-medium">Subscription</span>
+      <Button
+        variant="link"
+        size="sm"
+        className="h-auto p-0 text-sm"
+        onClick={() => void openPortal()}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply frontend/src/pages/billing/PlanDetailsCard.tsx]: Good catch — added toast error handling matching the pattern in DowngradeSection (`18b8e78`).

File: frontend/src/hooks/useBillingPortal.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+      }
+      return data;
+    },
+  });
+
+  return {
+    openPortal: async () => {
+      const result = await mutation.mutateAsync();
+      window.location.href = result.url;
+    },
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply frontend/src/hooks/useBillingPortal.ts:35]: Added `isStripeUrl` validation before redirect, matching the UpgradeCTA pattern (`070a8ca`).

File: api/billing_test.go [hunk 6]
```diff
@@ -185,6 +185,42 @@ func TestGetBillingPlan_PaidPlan(t *testing.T) {
...
+	var resp billingPlanResponse
+	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
+		t.Fatalf("failed to parse response: %v", err)
+	}
+	if resp.Subscription.HasPaymentMethod {
+		t.Error("expected has_payment_method=false for downgraded user with stale stripe_subscription_id")
+	}
+}
+
 // ── GET /billing/subscription ─────────────────────────────────────────────
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/billing_test.go:260]: Added both tests: `TestBillingPortal_RequiresStripeClient` (503) and `TestBillingPortal_NoCustomerID_Returns404` (404) (`070a8ca`).

File: api/billing_test.go [hunk 7]
```diff
@@ -978,3 +1014,47 @@ func TestListInvoices_NoSubscription(t *testing.T) {
...
+
+	r := authenticatedRequest(t, http.MethodPost, "/billing/portal", uid)
+	w := httptest.NewRecorder()
+	router.ServeHTTP(w, r)
+
+	if w.Code != http.StatusNotFound {
+		t.Fatalf("expected 404, got %d: %s", w.Code, w.Body.String())
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/billing_test.go:1096]: The codebase doesn't have a mock Stripe client — none of the other Stripe-dependent endpoints (checkout, upgrade, invoices) have happy-path tests either. Adding one here would require introducing a Stripe interface + mock, which is a larger refactor best done in its own PR. The error-path coverage we added is consistent with the existing test patterns.

File: api/billing_test.go [hunk 8]
```diff
@@ -185,6 +185,42 @@ func TestGetBillingPlan_PaidPlan(t *testing.T) {
...
+	}
+
+	var resp billingPlanResponse
+	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
+		t.Fatalf("failed to parse response: %v", err)
+	}
+	if resp.Subscription.HasPaymentMethod {
+		t.Error("expected has_payment_method=false for downgraded user with stale stripe_subscription_id")
+	}
+}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/billing_test.go:258]: Added `TestGetBillingPlan_HasPaymentMethodTrueViaStripeSubscription` — pay-as-you-go user with `stripe_subscription_id` set and zero local payment methods, asserts `has_payment_method=true` (`52e5b7b`).

**Comment by chiedoclaude** [comment]
## 🤖 Watch Session Summary

### Changes Made
No changes made during this session.

### Decision Log

#### ✅ Implemented
- greptile-apps[bot] asked for Fix false-positive HasPaymentMethod for downgraded users by gating on plan_id != free → implemented in `fd57738`
- greptile-apps[bot] asked for Add test coverage for downgraded user scenario → implemented in `fd57738`
- greptile-apps[bot] asked for Add error toast for billing portal failures instead of silently swallowing → implemented in `18b8e78`
- greptile-apps[bot] asked for Add isStripeUrl validation before portal redirect to prevent open redirect → implemented in `070a8ca`
- greptile-apps[bot] asked for Add RequiresStripeClient and NoCustomerID tests for portal endpoint → implemented in `070a8ca`
- greptile-apps[bot] asked for Add test for primary fix: paid user with stripe_subscription_id but no local payment methods → implemented in `52e5b7b`

#### ❌ Declined
- greptile-apps[bot] suggested Add happy-path test for billing portal endpoint → declined because No mock Stripe client exists in the codebase; none of the other Stripe endpoints have happy-path tests either.

---
*Watch session ended after 6 minutes of inactivity. Processed 7 comments across 6 poll cycles.*

**Comment by chiedoclaude** [comment]
⚠️ **Auto-merge failed.** The `--automerge` flag was set, but the merge could not be completed. Please merge manually.
